File: workers/python/smoke.py
import asyncio
import json

import grpc

import worker


async def main():
    listen = "127.0.0.1:50081"
    server = grpc.aio.server()
    worker.worker_pb2_grpc.add_WorkerServiceServicer_to_server(
        worker.WorkerService("python-smoke"),
        server,
    )
    server.add_insecure_port(listen)
    await server.start()

    async with grpc.aio.insecure_channel(listen) as channel:
        stub = worker.worker_pb2_grpc.WorkerServiceStub(channel)
        health = await stub.Health(worker.worker_pb2.HealthRequest())
        assert health.worker_id == "python-smoke"

        response = await stub.Execute(
            worker.worker_pb2.ExecuteRequest(
                request_id="smoke-1",
                workflow_name="wf",
                node_id="tool-1",
                operation="tool",
                target="echo",
                payload_json=json.dumps({"input": {"x": 1}}),
            )
        )
        assert response.ok
        decoded = json.loads(response.output_json)
        assert decoded["language"] == "python"

    await server.stop(0)
    print("python smoke ok")


if __name__ == "__main__":
    asyncio.run(main())

File: workers/typescript/smoke.ts
import * as grpc from "@grpc/grpc-js";
import * as protoLoader from "@grpc/proto-loader";
import path from "node:path";
import { fileURLToPath } from "node:url";

import { startWorker } from "./worker.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const protoPath = path.resolve(
  __dirname,
  "../../crates/simple-agents-workflow-workers/proto/worker.proto",
);

const packageDefinition = protoLoader.loadSync(protoPath, {
  keepCase: true,
  longs: String,
  enums: String,
  defaults: true,
});
const proto = grpc.loadPackageDefinition(packageDefinition) as any;

async function unaryCall(client: any, method: string, payload: any): Promise<any> {
  return new Promise((resolve, reject) => {
    client[method](payload, (error: Error | null, response: any) => {
      if (error) {
        reject(error);
        return;
      }
      resolve(response);
    });
  });
}

async function runSmoke() {
  const server = startWorker("127.0.0.1:50083");
  const client = new proto.workflow.worker.v1.WorkerService(
    "127.0.0.1:50083",
    grpc.credentials.createInsecure(),
  );

  const health = await unaryCall(client, "Health", {});
  if (health.worker_id !== "typescript-0") {
    throw new Error("unexpected worker_id");
  }

  const response = await unaryCall(client, "Execute", {
    request_id: "smoke-1",
    workflow_name: "wf",
    node_id: "node-1",
    operation: "tool",
    target: "echo",
    payload_json: JSON.stringify({ input: { x: 1 } }),
  });

  if (!response.ok) {
    throw new Error("expected ok execute response");
  }

  client.close();
  await new Promise<void>((resolve, reject) => {
    server.tryShutdown((error) => {
      if (error) {
        reject(error);
        return;
      }
      resolve();
    });
  });
  console.log("typescript smoke ok");
}

runSmoke().catch((error) => {
  console.error(error);
  process.exit(1);
});
